Docs: improve docs narrative around "ENSNode Plugins", "ENSDb Writers", and "ENSDb Readers". (#2227)
namehash/ensnode: docs/ensnode.io/src/content/docs/docs/integrate/integration-options/ensnode-plugins.mdx

---
title: ENSNode Plugins
description: How to customize data models for ENSNode, using the ENSNode Plugin specification.
---

import { LinkCard } from "@astrojs/starlight/components";

## What is an ENSNode Plugin?

> An **ENSNode Plugin** is an abstract specification that defines how a specific aspect of the ENS namespace is indexed into [ENSDb](/docs/services/ensdb). It is **not** an implementation — it is a standard that any [ENSDb Writer](/docs/services/ensdb/concepts/glossary#ensdb-writer) can implement.

Reviewer: ```suggestion
An **ENSNode Plugin** is an abstract specification that defines how onchain data relevant to ENS should be indexed into [ENSDb](/docs/services/ensdb). It is **not** an implementation — it is a standard that any [ENSDb Writer](/docs/services/ensdb/concepts/glossary#ensdb-writer) can implement and that any [ENSDb Reader](/docs/services/ensdb/concepts/glossary#ensdb-reader) can read.
```

Each ENSNode Plugin defines:

- **Name** — e.g. `unigraph`, `protocol-acceleration`, `subgraph`.

> - **Datasources** — which onchain contracts should be indexed, as a function of the ENS namespace.

Reviewer: ```suggestion
- **Datasources** — which onchain contracts should be indexed, as a function of an ENS namespace.
```

- **Dependency relationships** — other plugins that must (or must not) be activated concurrently.

> - **Indexed data model** — the tables, columns, and indexes that the plugin produces in ENSDb.

Reviewer: ```suggestion
- **Indexed data model** — the tables, columns, and indexes that implementations of the plugin must produce in ENSDb.
```

> - **Standards and invariants** — rules for how onchain events are translated into the indexed data model during indexing.

Reviewer: ```suggestion
- **Standards and invariants** — rules for how onchain events from the datasources are translated into the indexed data model during indexing.
```

> - **Versioning** — a version number that can be stored in ENSNode Metadata, to track which version of the plugin implementation is being used by ENSDb Writers and ENSDb Readers.

Reviewer: ```suggestion
- **Versioning** — a version number that can be stored in the ENSNode Metadata, to track which version of the plugin standard is being created by an ENSDb Writer and is available to ENSDb Readers.
```

## Why ENSNode Plugins Matter

> ENSNode Plugins are the key architectural piece that enables a decoupling between [ENSDb Writers](/docs/services/ensdb/concepts/glossary#ensdb-writer) and [ENSDb Readers](/docs/services/ensdb/concepts/glossary#ensdb-reader), so that anyone can create their own implementations of these. For example, any team working on indexing infrastructure can implement an ENSDb Writer and have it produce a standards-compliant ENSDb. On the other hand, any team working on products for ENS ecosystem can build their own standards-compliant ENSDb Reader and have it query ENSDb.

Reviewer: ```suggestion
ENSNode Plugins are a key architectural piece that enables a decoupling between [ENSDb Writers](/docs/services/ensdb/concepts/glossary#ensdb-writer) and [ENSDb Readers](/docs/services/ensdb/concepts/glossary#ensdb-reader), so that anyone can create their own implementations of these. For example, any team working on indexing infrastructure can implement an ENSDb Writer and have it produce a standards-compliant ENSDb. On the other hand, any team working on products or infrastructure in the ENS ecosystem can build their own standards-compliant ENSDb Reader and have it query ENSDb.
```

:::note[Interoperability between ENSDb Writers and Readers]

> When new ENSNode plugins are created, it becomes possible for the [ENS Omnigraph API](/docs/integrate/omnigraph) (or other APIs built by any ENSDb Reader) to unify the ability to query across ENSNode plugins in a single query. For example, you could query both ENS state from the `unigraph` and EFP state about followers from the `efp` plugin in a single GraphQL query.

Reviewer: ```suggestion
When new ENSNode plugins are created, it becomes possible for the [ENS Omnigraph API](/docs/integrate/omnigraph) (or other APIs built by any ENSDb Reader) to unify the ability to query across ENSNode plugins in a single query. For example, you could query both ENS state from the `unigraph` and EFP state about followers from the `efp` plugin in a single Omnigraph GraphQL query.
```

:::

> ## Core Plugins

Reviewer: ```suggestion
## Existing Plugins
```

> [ENSIndexer](/docs/services/ensindexer) is a reference implementation of an [ENSDb Writer](/docs/services/ensdb/concepts/glossary#ensdb-writer) that implements multiple core plugins following the ENSNode Plugins specification. The code for each plugin inside ENSIndexer is a reference implementation of that plugin's abstract specification. [The core plugins implemented in ENSIndexer today](https://github.com/namehash/ensnode/tree/main/apps/ensindexer/src/plugins) are:

Reviewer: ```suggestion
[ENSIndexer](/docs/services/ensindexer) is a reference implementation of an [ENSDb Writer](/docs/services/ensdb/concepts/glossary#ensdb-writer) that implements multiple existing ENSNode Plugin specifications. The code for each plugin inside ENSIndexer is a reference implementation of that plugin's abstract specification. [The ENSNode Plugins implemented in ENSIndexer today](https://github.com/namehash/ensnode/tree/main/apps/ensindexer/src/plugins) are:
```

:::note[Built by the NameHash team]
All core ENSNode Plugins are built by the NameHash team. [We welcome contributions and collaborations to expand the ecosystem of plugins](#build-your-own).
:::

> 

Reviewer: ```suggestion
```

| Plugin                  | Description                                                |
| ----------------------- | ---------------------------------------------------------- |
| `unigraph`              | Unified ENSv1 + ENSv2 data model with polymorphic entities |
| `protocol-acceleration` | Accelerated lookups for ENS resolution                     |
| `subgraph`              | Legacy ENS Subgraph-compatible data model                  |
| `basenames`             | Basenames (`.base.eth`) subname indexing                   |
| `lineanames`            | Lineanames (`.linea.eth`) subname indexing                 |
| `threedns`              | 3DNS (`.box`) name indexing                                |
| `registrars`            | Registration and renewal lifecycle tracking                |

> | `tokenscope`            | NFT tokenization and marketplace activity                  |

Reviewer: ```suggestion
| `tokenscope`            | NFT tokenization of ENS names                  |
```

> ## Community Plugins

Reviewer: ```suggestion
## New Plugins
```

### Under Development

> The `efp` plugin is [a new plugin currently under development](https://github.com/Quantumlyy/efpnode/tree/main/packages/ensnode-plugin-efp) by the [EthId team](https://ethid.org/):

Reviewer: ```suggestion
An `efp` plugin for the [Ethereum Follow Protocol](https://efp.app/) is [already under active development](https://github.com/Quantumlyy/efpnode/tree/main/packages/ensnode-plugin-efp) by the [EthId team](https://ethid.org/).
```

- [EFP (Ethereum Follow Protocol)](https://efp.app/) — onchain social graph protocol

> - [Grails](https://grails.app/) — by the EthId Foundation

Reviewer: ```suggestion
```

### Build Your Own

If you are interested in defining a new ENSNode plugin or implementing an existing one, reach out to the NameHash Labs team — we are happy to provide support and additional info. Join us on [Telegram](https://t.me/ensnode) for any questions.

## Related

<LinkCard
  title="ENSDb Writers"
  description="Learn how to build indexers that implement ENSNode plugins."
  href="/docs/integrate/integration-options/ensdb-writers"
/>

<LinkCard
  title="ENSDb Readers"

>   description="Learn how to build APIs that read from ENSNode plugins."

Reviewer: ```suggestion
  description="Learn how to build APIs and other services that read from ENSNode plugins."
```
